### DISCUSSION TITLE
Example setup when API is already returns a JWT token

### DISCUSSION THREAD

**Discussion by exocode** [nextauthjs/next-auth#2711] [Help]
### Question 💬

Hi there, I have an external Rails API which already handles JWT generation. I did not found an example which handles that case. All examples are assuming that Next-Auth handles generation of JWT tokens, and external APIs only returning user data, when correct credentials are sent.

I have an opened SO question, which is not answered a few days now, and I completly stuck with my problem. so I thought the "makers" of this awesome lib can help me in my "very special case". Maybe my approach is not correct or Next-Auth is not needed anyway.

Here is my current approach:

```js
providers: [
    CredentialsProvider({
      name: 'Email',
      credentials: {
        email: { label: "Email", type: "email", placeholder: "meine.email@domain.com" },
        password: {  label: "Passwort", type: "password" }
      },
      async authorize(credentials) {

        // The 'url' is pointing to a Rails API endpoint which returns a JWT Token
        const url = `${process.env.NEXT_PUBLIC_API_URL}/auth/login`;

        const res = await fetch(url, {
          method: 'POST',
          body: JSON.stringify(credentials),
          headers: {
            "Content-Type": "application/json" }
        })
        const user = await res.json()

        // If no error and we have user data, return it
        if (res.ok && user) {
          // I SAW EXAMPLES RETURNING {"email": "blah@tst.com"}
          return user // MY CONTENT {token: 'eyJhbGciOiJIUzI1NiJ9.eyJ1c2VyX2lkIjo0LCJyb2xl…0.OAGiwjj9O_NsH02lIjA2D4HYZkmTQ3_SqtKcVgaIul0'}

        }
        // Return null if user data could not be retrieved
        return null
      }
    })
  ]
}
```

This is what my Rails API JWT decoded response contains: 
My Rails API Token Content:
```json

{
  "user_id": 4,
  "email": "test@test.com",
  "roles": [
    "user"
  ],
  "exp": 1631096219
}
```
Do I have to decode my API token and reassamble that within the Provider.Credentials function?

*Do I have to "remap" the response somehow with these `callback`-functions?* (https://next-auth.js.org/v3/configuration/callbacks)

I hope you understand my problem. For completion: here is the StackOverflow question I refering to: https://stackoverflow.com/questions/69086448/next-auth-with-provider-credentials-how-to-implement-when-api-is-already-return

Many thanks in advance

### How to reproduce ☕️

Make a request to an API which returns that response

`{token: 'eyJhbGciOiJIUzI1NiJ9.eyJ1c2VyX2lkIjo0LCJyb2xlcyI6WyJ1c2VyIl0sImV4cCI6MTYzMTIzNjQ4MX0.YRZcgUZ3IacONYc3t1H6omNTUcBFpcMtcGPi7iCzmpo'}`

### Contributing 🙌🏽

No, I am afraid I cannot help regarding this

**Comment by balazsorban44** [comment]
In short, yes, I think you will have to decode the token for use. We generally advise against rolling your own auth solution, hence the Credentials provider flow is by design harder to use for special use-cases.

I am not familiar with Rails API, but if it has an OAuth/OpenID Connect API, you would be much better off with that, as we support it very nicely

Docs: https://next-auth.js.org/configuration/providers/oauth-provider#using-a-custom-provider

Another idea is that you return both the decoded and raw token as an object, and you only pass the raw token to requests to your APIs that need auth. You can have a look at the callbacks to understand how to make the token available from the front end.

**Reply by exocode** [reply]
Rails API could be extended to your needs. There are a lot of OAuth extensions out there, but I don't need them in my case.

> Another idea is that you return both the decoded and raw token as an object, and you only pass the raw token to requests to your APIs that need auth. You can have a look at the callbacks to understand how to make the token available from the front end.

I assume you are thinking of a solution like that (strong simplyfied)

```js
  callback:
   async jwt(token, user) {
      if (user) {
        token.accessToken = user.data.token,
        token.email = decodeSomehowJWT(user.data.token).email
      }
      return token
  },
```

When I use this approach is this "correct"?
Do I have to add that to the `session` callback to?

**Reply by balazsorban44** [reply]
it looks correct to me. one thing you would need the session callback for is to forward these values to the client (eg when calling useSession)
https://next-auth.js.org/configuration/callbacks#session-callback

**Reply by exocode** [reply]
Hi  @balazsorban44 again, 
maybe you can help me I struggle with my setup when using `getSession()` on a api request. it always returns `null`

here are all relevant parts of the code..


[...nextauth].js

```js
  jwt: {                                                                          // <<<<< These lines are added 
    secret: process.env.NEXTAUTH_SECRET             // <<<<< in progress of my investigation
  },                                                                                   // <<<<< 
  secret: process.env.NEXTAUTH_SECRET,                // <<<<< 
  callbacks: {
    // This callback is called whenever a JSON Web Token is created (i.e. at sign in)
    // or updated (i.e whenever a session is accessed in the client).
    // The returned value will be signed and optionally encrypted,
    // and it is stored in a cookie
    // The arguments user, account, profile and isNewUser are only passed the first time
    // this callback is called on a new session, after the user signs in.
    // In subsequent calls, only token will be available.
    async jwt({
      token,
      user,
      account,
      profile,
      isNewUser
    }) {
      // This user return by provider {} as you mentioned above MY CONTENT {token:}
      if (user) {
        if (user.token) {
          var token = {
            token: user.token,
            name: user.name,
            email: user.email,
          };
        }
      }

      return token;
    },
```



The request

```ts
await axios.get(`${process.env.NEXT_PUBLIC_URL}/api/account`).then(res => {
            const account = res.data.data 
            return account
        }).catch((err: any) => {
            console.log("err", err);
            return err
        });
```


the API endpoint

```ts
import { NextApiRequest, NextApiResponse } from 'next';
import Stripe from 'stripe';
import { getToken } from "next-auth/jwt"
import { getSession } from "next-auth/react"
// returns products created in Engagebay
export default async (req: NextApiRequest, res: NextApiResponse) => {
    const secret = process.env.NEXTAUTH_SECRET
    const token = await getToken({ // <<<<<<<<<<<<<< ALWAYS NULL when not using {raw: true}
        req, secret, encryption: true
    }) // <<<<<<<<<<<<<<<<<<<< when raw true, I get the regular token... 
    const session = await getSession({ req })   // <<<<<<<<<<<<<<  Always null 
    if (!token) {

        res.status(401).json({
            statusCode: 401,
            message: 'Unauthorized'
        });
        return;
    }
}
```

Is this intentional or I am using this wrong. 
Remember: I use a outside generated JWT token (It is generated on my Rails API

Do I have to implement "DECODE /  ENCODE" functions in [...nextauth.js
 ?

**Reply by exocode** [reply]
For now I've a workaround


```ts
import jwtDecode from 'jwt-decode';

export interface TokenContent { [key: string]: string; }

// decodes given JWT token
export const tokenContent = (token: string): TokenContent => {
    if (token) {
        return jwtDecode(token);
    }
    return {};
};

export const isValidToken = (authToken: string) => {
    const content = tokenContent(authToken);
    if (content && content.exp) {
        const expirationDate = +content.exp;
        const currentTime = Date.now() / 1000;
        return currentTime < expirationDate;
    }
    return false;
};

```


API endpoint
```ts
import { NextApiRequest, NextApiResponse } from 'next';
import Stripe from 'stripe';
import { getToken } from "next-auth/jwt"
import { getSession } from "next-auth/react"
import { isValidToken, tokenContent } from '../../../lib/auth';
// returns products created in Engagebay
export default async (req: NextApiRequest, res: NextApiResponse) => {
    const secret = process.env.NEXTAUTH_SECRET
    const token = await getToken({ req, secret, raw: true })

    const isValid = isValidToken(token)
    const session = await getSession({ req })
    if (!isValid) {

        res.status(401).json({
            statusCode: 401,
            message: 'Unauthorized'
        });
        return;
    }
    const content = tokenContent(token)

```


But somehow it does not "feel right" using NextAuth this way... 

Any suggestions

**Reply by LouisLecouturier** [reply]
Hi ! I know this discussion is very old but have you found any "easy" workaroud ? or do I still need to decode my API jwt and pass its info to the user object ?
